keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,575 @@
+have the feature gate off). Assuming that the feature remains alpha /
+disabled-by-default for at least 2 releases, this should ensure that
+all non-early-adopters get smooth upgrades and downgrades.
+
+For early adopters, or if we want to only have the feature be alpha
+for a single release, and not have completely smooth behavior for
+users with N-2 version skew, there are a few potential problems:
+
+#### Handling of the "`kubernetes`" Service
+
+kube-apiserver attempts to force the "`kubernetes`" Service definition
+to match its expectations on startup, but does not attempt to
+reconcile it after that point. Old/current apiservers always expect
+the Service to be single-stack.
+
+Thus, during an upgrade in which some (new) apiservers are trying to
+use the `DualStackAPIServer` feature gate, and some (old) apiservers
+predate the new feature, whichever apiserver started last will
+determine whether the Service is currently single or dual stack.
+
+In a simple upgrade or downgrade, this would result in the Service
+definition changing when the first apiserver was upgraded/downgraded,
+and then not changing after that. In a more complicated
+upgrade/downgrade scenario, there might be some flapping; eg, the
+first master upgrades and changes the Service to dual-stack, but then
+the second master reboots before upgrading, causing the old apiserver
+to restart and revert the Service definition back to its single-stack
+form, before then being upgraded and changing it back to dual-stack
+again.
+
+At no point would the Service be advertised as dual-stack when there
+were no dual-stack apiservers, so the Service ought to always be
+usable; it just may sometimes be single-stack when it _could_ have
+been dual-stack.
+
+(The one way that something could go wrong would be if a spurious
+apiserver were to start up after the upgrade/downgrade completed. Eg,
+after upgrading all apiservers to a dual-stack release, if an old
+single-stack-only apiserver were to start up, but then exit, it might
+spuriously change the Service back to single-stack before exiting.
+It's not clear why this would happen, and at any rate, it would be
+fixed as soon as any of the real apiservers restarted again.)
+
+#### Handling of "`kubernetes`" Endpoints and EndpointSlices
+
+For the "`kubernetes`" `Endpoints` (_not_ `EndpointSlice`), there
+should be no flapping; both old and new apiservers will agree at all
+times about what endpoints are available _in the cluster's primary IP
+family_, and thus will always agree on the contents of the
+single-stack `Endpoints` resource.
+
+For `EndpointSlice`, current versions of kube-apiserver create a
+single `EndpointSlice` object named "`kubernetes`", and assume that
+other apiservers will do the same; they don't look at any other
+`EndpointSlice` objects when reconciling. To create dual-stack
+EndpointSlices, a new apiserver would need to create two objects with
+different names. Even if new apiservers use the same name as the old
```

Comment:
should we provide more details on the new names for the secondary endpoint slice?
right now, there is only one, for compatibility I think it has to be the primary ip family, but how will be named the secondary?
```
kubectl get endpointslices
NAME         ADDRESSTYPE   PORTS   ENDPOINTS    AGE
kubernetes   IPv4          6443    172.18.0.2   21h
```